New node tool implementation

wonder-sk opened the pull request:
See https://github.com/qgis/QGIS-Enhancement-Proposals/issues/69 for more details.

wonder-sk committed:
Initial port of CAD-enabled node tool

nyalldawson commented on src/app/nodetool/qgsnodetool2.cpp:
```diff
@@ -0,0 +1,1393 @@
+  if ( match && match->hasVertex() && match->layer() && match->layer()->crs() == destLayer->crs() )
+  {
+    QgsFeature f;
+    QgsFeatureIterator fi = match->layer()->getFeatures( QgsFeatureRequest( match->featureId() ) );
```
No attributes

nyalldawson reviewed (commented):
Sounds great - can't wait to test!! 

I've posted a couple of suggestions with the code.

Two questions:
1. I see a lot of QgsPoint use here... Have you checked to see if z/m values are always retained here?
2. The old node tool went crazy if you deleted nodes in a multigeometry. Does this version handle this ok?

nyalldawson commented on src/app/nodetool/qgsnodetool2.h:
```diff
@@ -0,0 +1,260 @@
+    // members used for temporary highlight of stuff
+
+    //! marker of a snap match (if any) when dragging a vertex
+    QgsVertexMarker* mSnapMarker;
```
Oops... Bools!

nirvn commented:
@wonder-sk , nice. Will that resurrect snapping when moving existing nodes?

wonder-sk committed:
Fixed some bits from the review

wonder-sk committed:
Fix spelling

wonder-sk committed:
Fix build warning

wonder-sk committed:
Added unit testing for the new node tool

wonder-sk committed:
Fix rare crash on map canvas deletion

wonder-sk committed:
More unit tests

wonder-sk committed:
Code cleanup + more testing

wonder-sk commented on src/app/nodetool/qgsnodetool2.cpp:
```diff
@@ -0,0 +1,1393 @@
+
+QgsNodeTool2::~QgsNodeTool2()
+{
+  delete mSnapMarker;
```
I do not have a strong opinion here, but for pointers that exist for the whole lifetime of the node tool I do not see that much advantage of using unique pointers...

nyalldawson replied:
I see your point for a tool like this, but I think it's just a good habit for us to get into. Plus, we still want to be sure we don't leak even with objects which are bound to the lifetime of QGIS as it avoids valgrind noise.

wonder-sk commented:
Thanks a lot @nyalldawson for the review - I have fixed various bits based on your feedback... For your questions - I haven't really played much with geometries with z/m and with multi-geometries - will check later.

@nirvn yes that also resurrects snapping :-)

nyalldawson commented:
@wonder-sk I just gave a trial run for this. It's very nice!!

Here's a couple of issues I ran into (small stuff):
- the bounding box for the red dot doesn't seem big enough, and i get artefacts as it moves around
- if i have the "show markers only for selected features" option ticked, then i don't see the marker points when I'm editing a non-selected feature. The old node tool always shows these nodes regardless of whether a feature is selected.
- i can make a selection of multi vertices, but how do i move these?

pcav commented:
Martin, while you are at it, would you mind also changing the colour scheme? People find the current one (red for normal nodes, blue for selected, green for errors) quite counter intuitive. Probably a semaphore (gree for normal, yellow for selected, red for errors) would be OK.
Thanks.

wonder-sk commented:
> the bounding box for the red dot doesn't seem big enough, and i get artefacts as it moves around

Hmm - I do not really see any artifacts unfortunately... Which point "red dot" do you mean - highlight of the vertex that can be moved? The tool only uses existing classes QgsRubberBand and QgsVertexMarker for drawing things, so it must be a problem in one of those... Unfortunately I don't know how I could debug that (or maybe I just don't know how to replicate it) - could you maybe try to increase bounding box in those classes?

> if i have the "show markers only for selected features" option ticked, then i don't see the marker points when I'm editing a non-selected feature. The old node tool always shows these nodes regardless of whether a feature is selected.

Actually that option is ignored by the old node tool - it always highlights the feature that you click on. The "show markers only for selected features" really only kicks in if you use select tool and select some features.

In my opinion, the way how editing markers are shown should be different. By default no markers should be shown, and there could be a button in the toolbar to enable/disable display of markers...

> i can make a selection of multi vertices, but how do i move these?

You cannot currently - but you can delete them all at once :-) It is on my to-do list...

wonder-sk commented:
> Martin, while you are at it, would you mind also changing the colour scheme? People find the current one (red for normal nodes, blue for selected, green for errors) quite counter intuitive. Probably a semaphore (gree for normal, yellow for selected, red for errors) would be OK.

@pcav I agree it would be good to address that, but I would prefer to address that separately later rather than on this PR...

wonder-sk committed:
Trying to understand what's wrong on travis...

pcav commented:
@wonder-sk ok, so how do you suggest to best proceed not to forget? Should I open a ticket about this?

wonder-sk committed:
More travis debugging

wonder-sk commented:
@pcav yes feel free to open a ticket and suggest different colors

wonder-sk committed:
More travis debugging

nyalldawson commented:
@wonder-sk here's what I see: https://www.youtube.com/watch?v=SjnLK8z-4VE&feature=youtu.be

wonder-sk committed:
More travis debugging

wonder-sk committed:
More travis testing

wonder-sk committed:
travis

wonder-sk committed:
Looks like crs mismatch...?

pcav commented:
Done, thanks: http://hub.qgis.org/issues/16260

wonder-sk committed:
fixes after rebase

wonder-sk committed:
Fix rubber band bounding box calculation

The icon half-width was incorrectly divided by map units per pixel...

wonder-sk commented:
@nyalldawson hopefully 2e0df54 should solve your issues with markers

wonder-sk committed:
Fix rubber band unit tests

wonder-sk committed:
Double-click to add vertex only on an edge

wonder-sk committed:
Do not leave rubber band artifacts when deleting vertices

wonder-sk committed:
Fix highlight of marker to add vertices to linestrings

wonder-sk committed:
Remove debugging output to track down travis failure

wonder-sk committed:
Added test for dragging of selection rect to delete nodes

wonder-sk committed:
Little bit of renaming of variables to camelCase

wonder-sk committed:
Support for moving of multiple vertices at the same time

Just use selection rectangle to select vertices and then click one vertex
from selection to move the whole group.

wonder-sk commented:
Now the new node tool is able to move multiple vertices at the same time :-)

wonder-sk committed:
Move private arc segmentization routines to QgsGeometryUtils

wonder-sk committed:
Correctly show rubber bands when dragging vertices of circular edges

wonder-sk committed:
Simplify code for straight line rubber bands

nirvn commented:
@wonder-sk , nice to see we're approaching the day this will be committed to master :)

wonder-sk committed:
Simplify code for dragging of edges

wonder-sk committed:
Fix moving of edges with circular edges + code cleanups

wonder-sk committed:
Fix support for advanced digitizing dock (CAD)

wonder-sk committed:
Fix topological editing when snapping a node to a segment

wonder-sk committed:
Fixed display of rubber bands when using OTF reprojection

wonder-sk committed:
Slightly improved markers for highlighted nodes and circular nodes

wonder-sk committed:
Integrate node editor from the old node tool

When right-clicking on a feature, a popup menu will be shown
and it is possible to open node editor dock widget for that particular
feature to edit coordinates numerically or to delete vertices.

wonder-sk committed:
Update highlighted nodes when features are changed or deleted

wonder-sk committed:
Add validation of changed geometries

Whenever geometries are modified, validation will be run while node tool
is active and validation errors are displayed in status bar and with markers on canvas

wonder-sk committed:
Fix build for Qt<5.6

wonder-sk committed:
Spelling fix

wonder-sk committed:
Also highlight vertices when highlighting a feature on mouse hover

wonder-sk committed:
Fix snapping configuration to snap only to editable layers

wonder-sk committed:
Cleanup of validation objects and markers

wonder-sk commented:
I think the new node tool is ready for final user testing, review and merge. I would like to merge it early next week if there will be no objections.

This PR does not remove the old node tool and old snapping classes yet - the removal will be done in a follow up PR.

nyalldawson commented:
Soooo nice - this is a huge improvement! :D 

A couple of small issues I've noted:
- using < > doesn't wrap around polygons anymore - in the older node tool it'd jump back/forward when hitting the first/last node
- on the old node tool using < > caused canvas to "follow" the selected node and recenter the canvas if the node isn't visible in the current extent
- if i try to move a point geometry using the node tool and the advanced digitizing panel and restrict the x or y, things go crazy and the point disappears somewhere. This only happens if I change the x/y value - if i leave them untouched and lock then it works OK
- if i select a node with the advanced digitizing dock open, the initial x and y aren't populated - i have to move the node before the values are shown. The initial value should be shown so that it can be locked or refined from the initial value.
- things go a bit crazy if i try to use the advanced digitizing constraints while moving a segment rather than a node

Otherwise seems great to me, fantastic work!

wonder-sk committed:
Fixes for moving between nodes with keyboard

- supports < and > in addition to comma and dot
- wraps around polygon rings (jumps from last node to the first)
- moves view to node when it is outside of canvas view

wonder-sk committed:
Move from "or", "and" to proper C-style ||, &&

wonder-sk committed:
Fix map canvas destruction fix (got broken after rebase)

wonder-sk commented:
Thanks a lot @nyalldawson for testing...

> - using < > doesn't wrap around polygons anymore - in the older node tool it'd jump back/forward when hitting the first/last node
> - on the old node tool using < > caused canvas to "follow" the selected node and recenter the canvas if the node isn't visible in the current extent

Should be fixed now.

> - if i try to move a point geometry using the node tool and the advanced digitizing panel and restrict the x or y, things go crazy and the point disappears somewhere. This only happens if I change the x/y value - if i leave them untouched and lock then it works OK
> - if i select a node with the advanced digitizing dock open, the initial x and y aren't populated - i have to move the node before the values are shown. The initial value should be shown so that it can be locked or refined from the initial value.
> - things go a bit crazy if i try to use the advanced digitizing constraints while moving a segment rather than a node

I can replicate the second problem, but unfortunately I can't replicate the first and third one. All of these are however related to advanced digitizing widget and not directly related to node tool I think. In any case, I should have time to do a bit of refactoring of advanced digitizing dock widget - to separate logic from GUI, add unit tests and mainly clean up the API (reverse the flow - rather than map tools accepting whatever the dock widget comes up with (which works fine with simple tools, but not complex ones like node tool), map tools should be able to dictate how the advanced digitizing dock behaves). So if that's fine, I would defer those issues to a later time when I get to the refactoring (should be hopefully in a couple of weeks)

nirvn commented:
Wouhou!

3nids commented:
awesome @wonder-sk !
Finally this lands in QGIS!

From a simple first try out, I have two remarks:
* isn't double-clicking on a segment supposed to place a new node on that position?
* is moving of several selected nodes possible? how? can this be achieved on different layers/features at the same time?

wonder-sk commented:
> isn't double-clicking on a segment supposed to place a new node on that position?

Yes - and it also lets you to place the new node where you need it... doesn't it work for you that way?

> is moving of several selected nodes possible? how? can this be achieved on different layers/features at the same time?

Yes it is possible - just select multiple nodes (by dragging a rectangle) and then click one of the nodes to start moving the whole group. The node you clicked will be the "leader" that can be snapped or modified in CAD dock widget.

And yes it works with different features / layers at the same time!

3nids commented:
> Yes - and it also lets you to place the new node where you need it... doesn't it work for you that way?

ok, right. But I thought it would place the vertex and leave it in place (no move allowed). No big deal, just a misunderstanding.

> Yes it is possible - just select multiple nodes (by dragging a rectangle) and then click one of the nodes to start moving the whole group. The node you clicked will be the "leader" that can be snapped or modified in CAD dock widget.

Perfect. It might not be very intuitive for this one. But it works perffectly.

Thanks again for this!!!!

nyalldawson commented:
@wonder-sk @3nids 

I've been using this new node tool a lot today. One small enhancement idea:

I find myself always right clicking to end the node moving (I guess because other map tools use right click to "lock in" the action). This just cancels the action though. Do you think right click while an operation is being performed should act the same as the left click?

nirvn commented:
@wonder-sk , while we're at it, I noticed yesterday that dragging a segment ignores snapping settings.

wonder-sk commented:
@nyalldawson Hmm... a difficult question :-) If we use right-click to confirm the move, how do we do cancellation - with Esc key only? For me the left-click to accept and right-click to cancel feel quite natural - but I do not have a very strong opinion on that as I do not do heaps of vector editing...

@nirvn That is actually intentional as we can't really tell which endpoint should be the one that gets snapped... or do you think that both endpoints should be checked for snapping?

nirvn commented:
@wonder-sk , IMHO, yeah, I think both endpoints should be checked for snapping.

joaodiasafonso commented:
Hello. 
@wonder-sk : does the creation of a new vertex in a 3D geometry now gets an interpolated Z value, like how it happens in tools like reshape, or is still set to 0 (qgis bug report #16240)?
Thanks in advance! Regards,

wonder-sk commented:
@joaodiasafonso the bug is still there - when fixed the original bug report will be closed.

joaodiasafonso commented:
Ok! Considering that bug report date, this thread date and the last qgis release date, thought it could be implemented already (well, mostly because this thread was about a "new node tooI"...).
Is there any predictable deadline for it to be fixed?

gioman commented:
> Ok! Considering that bug report date, this thread date and the last qgis release date, thought it could be implemented already (well, mostly because this thread was about a "new node tooI"...).
> Is there any predictable deadline for it to be fixed?

@joaodiasafonso a way to make things move quicker is to support the development/fixing of critical issues. Cheers!

joaodiasafonso commented:
Thanks for the comment, Giovanni. I'm not trying to push anyone to do anything, just trying to figure out if someone is doing anything at all, specially considering this recent effort. The entry point/learning curve for qgis core is quite "expensive" at the moment for me, and I believe that a new plugin for that is just reinventing the wheel. Though if there's not a clue about further developments on this, I'll go that way.

gioman commented:
@joaodiasafonso probably I explained myself not very well. I just meant that is not unusual (nowadays is really common) for an entity that made a choice (in this case would be using QGIS) to accelerate the process of fixing a mission critical bug/void by supporting the work or even by submitting patched. I didn't meant that could be the case. Recently a 30000 euros crowdfunding was successful (with a part of it supported by several national users groups, Portugal included) and will improve the composer and add reporting capabilities to QGIS. Maybe the same could be done for this bug? I think that the possible values involved could be pretty different (much less).

joaodiasafonso commented:
Ok! Now I fully understand your "support" suggestion:) I haven't considered such alternatives... Thanks for the tip!

m-kuhn commented:
@wonder-sk what do you think about moving the new node tool over to the gui library?

Sometimes it would be handy to implement some derived tools based on this class.

wonder-sk commented:
@m-kuhn I would not mind, however it does not currently provide means for derived classes to customize the behavior. That said, if you feel like you could have a use for it, feel free to move it to GUI lib...

m-kuhn commented:
One could at least always override the click and move events. Anything more can be inserted down the road as requirements arise.
The main question here is "is the implementation clean enough?" And if yes, let's go :)

wonder-sk commented:
> One could at least always override the click and move events.

Right but that's somehow limited :-) Do you have a particular use case in mind?

> The main question here is "is the implementation clean enough?"

It is the most beautiful piece of code the world has ever seen :stuck_out_tongue:

m-kuhn commented:
> > One could at least always override the click and move events.
> 
> Right but that's somehow limited :-) Do you have a particular use case in mind?

The requirement for one tool was "if the plugin is in some mode and a geometry gets modified, don't update the original geometry but create a modified copy". I think it's now using your python implementation of the node tool for 2. 18 :)

> > The main question here is "is the implementation clean enough?"
> 
> It is the most beautiful piece of code the world has ever seen :stuck_out_tongue:

Holy fucking moly!! :tada:

NathanW2 commented:
It's true. I have seen Martin's code. Pretty much God like :p

On Wed., 20 Sep. 2017, 7:58 pm Matthias Kuhn <notifications@github.com>
wrote:

> One could at least always override the click and move events.
>
> Right but that's somehow limited :-) Do you have a particular use case in
> mind?
>
> The requirement for one tool was "if the plugin is in some mode and a
> geometry gets modified, don't update the original geometry but create a
> modified copy". I think it's now using your python implementation of the
> node tool for 2. 18 :)
>
> The main question here is "is the implementation clean enough?"
>
> It is the most beautiful piece of code the world has ever seen 😛
>
> Holy fucking moly!! 🎉
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4187#issuecomment-330805094>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3DJk2AtZZTg6LK2SXutPqNDXQqLNks5skOGxgaJpZM4ML8oT>
> .
>
